feat(client) ✨: Add Role Receival

SBAsai opened the pull request:
This pr will add the role receiving component to the game. Upon reaching the component, the player will receive a role at random (CAT or CATFISH) to which they are granted different objectives and different prompts after the component is over.

nstringham committed:
feat(client): ✨ Cherrypick Private Player Collection

cmcclusky committed:
feat(client): ✨ Create structure for receive role.

Created a structure for the assignRole funtion, going to add documentation and fix Player errors.

cmcclusky committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into receive-role

cmcclusky committed:
feat(client): ✨ Complete function assignRole.

Complete the function assignRole, still needs to be tested with soon to be created UI.

SBAsai committed:
feat(client): ✨ Add basic css to page

Added a basic css layout to test the role retreival function

cmcclusky committed:
Merge branch 'development' into receive-role

cmcclusky committed:
Got assign role to function.

SBAsai committed:
feat(client): ✨ Make color-coded font color for role

Makes the text green for cat and red for catfish when appearing on screen

SBAsai committed:
Merge branch 'development' into asai-mcclusky/receive-role

SBAsai committed:
style(client): 🎨 Remove unused import, merge dev into branch

Removed setDoc from svelte file and merged dev into this branch

cmcclusky committed:
feat(client): ✨ Add changes from review, made ROLE a gamestate.

Added ROLE as game state for the time, unsure if we will keep it that way, more discussion tomorrow.

cmcclusky committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into asai-mcclusky/receive-role

SBAsai committed:
feat(client): ✨ Add description upon receiving role

Added a description that appears under where the role is shown (with puns) and updated the image to fit the theme better.

SBAsai committed:
Merge branch 'development' into asai-mcclusky/receive-role

cmcclusky committed:
Merge branch 'asai-mcclusky/receive-role' of https://github.com/cat-acronym-team/cat-me-if-you-can into asai-mcclusky/receive-role

cmcclusky committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into asai-mcclusky/receive-role

cmcclusky committed:
feat(client): ✨ Update to dev and attempt function

Pulled from development and attempted to assign roles from firebase functions.

nstringham reviewed (changes requested):
please resolve failing checks

nstringham commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,104 @@
+      </div>
+      <p class="description">Use your inner Purrlock Holmes and find the catfish!</p>
+      <!-- TODO: Use different cat picture -->
+      <img
+        src="https://media.istockphoto.com/photos/front-view-of-british-shorthair-cat-7-months-old-sitting-picture-id104355461?k=20&m=104355461&s=612x612&w=0&h=9KPbuZ6tTUxqi-_WZT8zrjNrf1W8XBuXIw8ZIM_YQuI="
+        alt="cat"
+      />
```
do we have rights to this image?

SBAsai replied:
You'll love the new images

nstringham commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,104 @@
+      <img
+        src="https://media.istockphoto.com/photos/channel-catfish-picture-id503331450?k=20&m=503331450&s=612x612&w=0&h=_ciR7YrmZN1EIooeMl1mo0aVJ3bGTXGVoFzBWVlTyts="
+        alt="catfish"
+      />
```
do we have rights to this image?

nstringham commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,104 @@
+  import { authStore as user } from "$stores/auth";
+
+  export let lobbyCode: string;
+  let getRole: PrivatePlayer;
```
there is a privatePlayer in `game/+page.svelte` so this can be an export let

nstringham commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,104 @@
+
+  export let lobbyCode: string;
+  let getRole: PrivatePlayer;
+  alert("Hello!");
```
what is this alert for?

nstringham commented on client/src/lib/firebase/role.ts:
```diff
@@ -0,0 +1,43 @@
+import { getPrivatePlayerCollection } from "$lib/firebase/firestore-collections";
+import { doc, setDoc, getDoc, DocumentReference } from "firebase/firestore";
+
+export async function assignRole(lobby: DocumentReference<Lobby>) {
```
this should not be on the client side

nstringham commented on client/src/lib/firebase/vote.ts:
```diff
@@ -0,0 +1,41 @@
+import { lobbyCollection } from "$lib/firebase/firestore-collections";
+import { getVoteCollection } from "$lib/firebase/firestore-collections";
+
+export async function vote(person: number, lobbyData: Lobby, code: string, uid: string) {
```
why is this part of this PR?

SBAsai replied:
I'm guessing Cameron forgot to switch branches, I'm gonna delete this file

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -99,6 +100,8 @@
   {:else if lobby.state === "CHAT"}
     <ChatRoom lobbyData={{ ...lobby, id: lobbyCode }} />
+  {:else if lobby.state === "ROLE"}
+    <Role {lobbyCode} />
```
lets put these in the order that they will happen in the game so "ROLE" should be before "PROMPT"

nstringham commented on functions/src/lobby.ts:
```diff
@@ -195,6 +198,32 @@ function startPrompt(lobbyDocRef: firestore.DocumentReference<Lobby>) {
   });
 }
 
+function setRole(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  const numCatFish = 1;
+  const check: number[] = [];
+  let current = 0;
+  const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+
+  return db.runTransaction(async (transaction) => {
+    const lobbyDoc = await transaction.get(lobbyDocRef);
+
+    const lobbyData = lobbyDoc.data();
+
+    for (let i = 0; i < numCatFish; i++) {
+      current = Math.floor(Math.random() * (lobbyData.uids.length - 1) + 0);
+      if (check.indexOf(current) === -1) {
+        check.push(current);
+      } else i--;
+    }
+
+    for (let j = 0; j < lobbyData.uids.length; j++) {
+      const privatePlayerDocRef = privatePlayerCollection.doc(lobbyData.uids[j]);
+
+      transaction.set(privatePlayerDocRef, { role: check.indexOf(j) === -1 ? "CAT" : "CATFISH" }, { merge: true });
+    }
+  });
+}
+
```
this code is really hard to read. I can barly figure out what is going on maybe consider more desriptive variable names

SBAsai replied:
I changed it up a bit at put it in `role.ts`. Let me know if you can understand it then

nstringham commented on functions/src/lobby.ts:
```diff
@@ -195,6 +198,32 @@ function startPrompt(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+    for (let j = 0; j < lobbyData.uids.length; j++) {
+      const privatePlayerDocRef = privatePlayerCollection.doc(lobbyData.uids[j]);
+
+      transaction.set(privatePlayerDocRef, { role: check.indexOf(j) === -1 ? "CAT" : "CATFISH" }, { merge: true });
```
maybe consider [Array.includes()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes)

nstringham commented on functions/src/lobby.ts:
```diff
@@ -149,6 +149,9 @@ export const onLobbyUpdate = functions.firestore.document("/lobbies/{code}").onU
     await startPrompt(lobbyDocRef);
   }
+  if (lobby.state == "ROLE" && oldLobby.state != "ROLE") {
+    await setRole(lobbyDocRef);
```
this should probobly be inside `startGame`

nstringham reviewed (dismissed):
didn't look over everything very closly but this should be a good starting point

SBAsai committed:
fix(functions): 🎨 Fix role assigning function to properly assign roles in server-side and add pictures.

Moved the role assigning function to the server-side. Added exquisite pictures that will be seen upon receiving a role. Moved function call to the startgame function in lobby. Fixed issues in game page.

SBAsai committed:
Merge branch 'development' into asai-mcclusky/receive-role

SBAsai committed:
style(client): 🎨 Fix css for image sizing

Added a container around the image to sizing purposes

nstringham commented on functions/src/lobby.ts:
```diff
@@ -87,12 +88,8 @@ export const startGame = functions.https.onCall(async (data: unknown, context):
-  }
-
-  await lobbyCollection.doc(data.code).update({ state: "PROMPT" });
+  await lobbyCollection.doc(data.code).update({ state: "ROLE" });
```
should move this inside the transaction in assignRole to prevent the problem we were having earlyer today where multiple states get skiped

nstringham committed:
refactor(functions): ♻️ use a set to streamline assignRole code

nstringham committed:
refactor(functions): ♻️ use create instead of set for initalizing privatePlayer

becasue set garentees that the type is correct

nstringham committed:
refactor(functions): ♻️ use for loop instead of primose all map

because the loop doesn't contain any awaits

nstringham committed:
docs(functions): 💡 update comment

nstringham commented on functions/src/role.ts:
```diff
@@ -0,0 +1,50 @@
+import { db } from "./app";
+import { firestore } from "firebase-admin";
+
+export function assignRole(lobby: firestore.DocumentReference<Lobby>) {
```
I had some ideas for how to improve this so I created #115

gchamb replied:
pass in the transaction as a parameter

gchamb replied:
pass in DocumentSnapshot instead of Document Reference

nstringham reviewed (dismissed):
everyone like the ears so we should remove the catfish without ears and rename the one with ears to just catfish

SBAsai committed:
Merge pull request #115 from cat-acronym-team/nstringham/receive-role

refactor(functions): ♻️ receive role

gchamb commented:
pull from dev

SBAsai committed:
style(client): 🎨 Change role receiving function structure. Remove cat ear-less catfish png

SBAsai committed:
Merge branch 'development' into asai-mcclusky/receive-role

gchamb commented on functions/src/lobby.ts:
```diff
@@ -90,13 +91,7 @@ export const startGame = functions.https.onCall(async (data: unknown, context):
-
-    // await lobbyCollection.doc(data.code).update({ state: "PROMPT" });
-    await startPrompt(lobby, transaction);
+    await assignRole(lobby.ref);
```
this function times out because you're using a transaction within a transaction. So it never goes to the ROLE state

gchamb commented on functions/src/role.ts:
```diff
@@ -0,0 +1,44 @@
+      } else {
+        transaction.create(privatePlayerDocRef, { role: "CAT" });
+      }
+      transaction.update(lobby, { state: "ROLE" });
```
you probably dont wanna update the the lobby state to ROLE multiple times

SBAsai replied:
I'm assuming you mean I should take it out of the for-loop

gchamb commented on functions/src/role.ts:
```diff
@@ -0,0 +1,44 @@
+  // Number of catfish below, will later be made to be changable by users.
+  const numCatFish = 1;
+
+  return db.runTransaction(async (transaction) => {
```
use the transaction from the function you call this from

gchamb reviewed (dismissed):
see above

SBAsai committed:
style(functions): 🐛 Fix transaction inside another transaction call. Move game state update to outside loop.

SBAsai committed:
Merge branch 'development' into asai-mcclusky/receive-role

SBAsai committed:
fix(functions): 🐛 Fix build error

Fixed build functions error by setting stalker to false upon receiving role

nstringham commented on functions/src/role.ts:
```diff
@@ -0,0 +1,41 @@
+    }
+  }
+  transaction.update(lobby, { state: "ROLE" });
+  // TODO: Change game state to "PROMPT" with timer
```
timer has bean merged now so you can do this

nstringham committed:
docs(functions): ✏️ fix spelling typo

nstringham commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,88 @@
+  import type { PrivatePlayer } from "$lib/firebase/firestore-types/lobby";
+
+  export let privatePlayer: PrivatePlayer;
+</script>
```
this is where the timers should go. you can look at `Result.svelte` if you need an example

SBAsai committed:
feat(client): ✨ Add timer to role receival page

Added a functional timer to the receive role page to continue to the prompt page

gchamb commented on client/src/components/ReceiveRole.svelte:
```diff
@@ -0,0 +1,124 @@
+
+  let countdown = GAME_STATE_DURATIONS.VOTE;
+  let timer: ReturnType<typeof setInterval>;
+  let votedOffUser: Player | undefined;
```
not used variable

SBAsai replied:
I noticed that I forgot to delete the voted off stuff too

gchamb commented on functions/src/role.ts:
```diff
@@ -0,0 +1,45 @@
+  // Number of catfish below, will later be made to be changeable by users.
+  const numCatFish = 1;
+  const lobby = lobbySnap.ref;
+  const validLobby = await transaction.get(lobby);
```
dont need this anymore since you have the lobbySnap as a parameter

gchamb commented on functions/src/role.ts:
```diff
@@ -0,0 +1,45 @@
+  const numCatFish = 1;
+  const lobby = lobbySnap.ref;
+  const validLobby = await transaction.get(lobby);
+  const lobbyData = validLobby.data();
```
refactor into to 
```ts 

const lobbyData = lobbySnap.data();

```

gchamb reviewed (dismissed):
see above after that everything should be good

SBAsai committed:
refactor(client): ♻️ Remove unused code

Removed unused code in the client. Refactored getting data from a lobby document snapshot

gchamb commented on functions/src/lobby.ts:
```diff
@@ -90,13 +91,7 @@ export const startGame = functions.https.onCall(async (data: unknown, context):
-
-    // await lobbyCollection.doc(data.code).update({ state: "PROMPT" });
-    await startPrompt(lobby, transaction);
+    await assignRole(lobby, transaction);
```
remove the await

SBAsai committed:
refactor(functions): 🎨 Remove await

Removed await from a function call
